p2p/transport/tcp/conn/connection.go

```diff
@@ -116,8 +98,16 @@ type MConnection struct {
 	created time.Time // time of creation
 
 	_maxPacketMsgSize int
+
+	mtx sync.RWMutex
+	// streamID -> list of incoming messages
+	recvMsgsByStreamID map[byte]chan []byte
```

Comment:
error prone + "unwanted" queue => this is why I don't think it shouldn't be merged into `main`. The only way is to remove it and bring back old API like `onReceive(...)`